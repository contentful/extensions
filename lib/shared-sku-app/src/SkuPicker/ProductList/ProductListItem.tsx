import React, { useState } from 'react';
import tokens from '@contentful/forma-36-tokens';
import noop from 'lodash/noop';
import { Icon, SkeletonContainer, SkeletonImage } from '@contentful/forma-36-react-components';
import { css } from 'emotion';
import { activeProductCheck } from '../iconsInBase64';
import { Product } from '../../interfaces';

export interface Props {
  product: Product;
  selectProduct: (sku: string) => void;
  isSelected: boolean;
}

const styles = {
  productWrapper: css({
    display: 'flex',
    flex: `0 0 calc(25% - ${parseFloat(tokens.spacingS) * 2}rem)`,
    padding: tokens.spacingS,
    position: 'relative'
  }),
  product: css({
    border: '1px solid',
    borderColor: tokens.colorElementLight,
    borderRadius: '3px',
    boxShadow: '0px 0px 0px 1px inset rgba(48, 114, 190, 0), 0 1px 3px rgba(0, 0, 0, 0)',
    display: 'flex',
    flexDirection: 'column',
    padding: tokens.spacingS,
    outline: 0,
    transition: `all ${tokens.transitionDurationDefault} ${tokens.transitionEasingDefault}`,
    '&:hover': {
      borderColor: tokens.colorElementDark,
      cursor: 'pointer'
    },
    width: '100%',
    // Force hardware acceleration for non-accelerated
    // animated props box-shadow and border-color
    transform: 'translateZ(0)',
    willChange: 'box-shadow, border-color'
  }),
  selectedProduct: css({
    borderColor: 'rgba(48, 114, 190, 1)',
    boxShadow: '0px 0px 0px 1px inset rgba(48, 114, 190, 1), 0 1px 3px rgba(0, 0, 0, 0.08)',
    '&:hover': {
      borderColor: 'rgba(48, 114, 190, 1)',
      cursor: 'pointer'
    }
  }),
  imgWrapper: (imageHasLoaded: boolean) =>
    css({
      width: '290px',
      height: `${imageHasLoaded ? 290 : 0}px`,
      position: 'relative',
      overflow: 'hidden'
    }),
  previewImg: css({
    height: '290px',
    minWidth: 'auto',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  }),
  name: css({
    flex: '1 0 auto',
    fontWeight: 'bold',
    textTransform: 'capitalize'
  }),
  sku: css({
    flex: '0 1 auto',
    color: tokens.colorTextLight,
    fontSize: tokens.fontSizeS,
    marginTop: 0,
    marginBottom: 0
  }),
  skeletonImage: css({
    width: '100%',
    height: '290px'
  }),
  errorImage: css({
    backgroundColor: tokens.colorElementLightest,
    width: '100%',
    height: '290px',
    position: 'relative',
    svg: css({
      fill: tokens.colorTextLight,
      width: '100%',
      height: '50%',
      position: 'absolute',
      top: '50%',
      left: '50%',
      transform: 'translate(-50%, -50%)'
    })
  }),
  check: (isSelected: boolean) =>
    css({
      opacity: isSelected ? 1 : 0,
      position: 'absolute',
      top: tokens.spacingL,
      right: tokens.spacingL,
      transition: `opacity ${tokens.transitionDurationDefault} ${tokens.transitionEasingDefault}`
    })
};

export const ProductListItem = (props: Props) => {
  const { product, isSelected, selectProduct } = props;
  const [imageHasLoaded, setImageHasLoaded] = useState(false);
  const [imageHasErrored, setImageHasErrored] = useState(false);

  return (
    <div className={styles.productWrapper}>
      <div
        data-test-id={`product-preview-${product.sku}`}
        role="switch"
        aria-checked={isSelected}
        tabIndex={-1}
        className={`${styles.product} ${isSelected ? styles.selectedProduct : ''}`}
        onKeyUp={noop}
        onClick={() => selectProduct(product.sku)}>
        {!imageHasLoaded && !imageHasErrored && (
          <SkeletonContainer className={styles.skeletonImage}>
            <SkeletonImage width={400} height={290} />
          </SkeletonContainer>
        )}
        {imageHasErrored && (
          <div className={styles.errorImage}>
            <Icon icon="Asset" />
          </div>
        )}
        {!imageHasErrored && (
          <div className={styles.imgWrapper(imageHasLoaded)}>
            <img
              onLoad={() => setImageHasLoaded(true)}
              onError={() => setImageHasErrored(true)}
              style={{ display: imageHasLoaded ? 'block' : 'none' }}
              src={product.image}
              alt="product preview"
              className={styles.previewImg}
              data-test-id="image"
            />
            <img className={styles.check(isSelected)} src={activeProductCheck} alt="check" />
          </div>
        )}
        <p className={styles.name}>{product.name}</p>
        <p className={styles.sku}>{product.sku}</p>
      </div>
    </div>
  );
};
