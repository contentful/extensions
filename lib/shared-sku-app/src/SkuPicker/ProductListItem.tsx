import React, { useState } from 'react';
import tokens from '@contentful/forma-36-tokens';
import noop from 'lodash/noop';
import { SkeletonContainer, SkeletonImage } from '@contentful/forma-36-react-components';
import { css } from 'emotion';
import { activeProductCheck } from './iconsInBase64';
import { Product } from '../interfaces';

export interface Props {
  product: Product;
  selectProduct: (sku: string) => void;
  isSelected: boolean;
}

const styles = {
  productWrapper: css({
    display: 'flex',
    flex: `0 0 calc(25% - ${parseFloat(tokens.spacingS) * 2}rem)`,
    padding: tokens.spacingS,
    position: 'relative'
  }),
  product: css({
    border: '1px solid',
    borderColor: tokens.colorElementLight,
    borderRadius: '3px',
    boxShadow: '0px 0px 0px 1px inset rgba(48, 114, 190, 0), 0 1px 3px rgba(0, 0, 0, 0)',
    display: 'flex',
    flexDirection: 'column',
    padding: tokens.spacingS,
    outline: 0,
    // textAlign: 'center',
    transition: `all ${tokens.transitionDurationDefault} ${tokens.transitionEasingDefault}`,
    '&:hover': {
      borderColor: tokens.colorElementDark,
      cursor: 'pointer'
    },
    width: '100%',
    // Force hardware acceleration for non-accelerated
    // animated props box-shadow and border-color
    transform: 'translateZ(0)',
    willChange: 'box-shadow, border-color'
  }),
  selectedProduct: css({
    borderColor: 'rgba(48, 114, 190, 1)',
    boxShadow: '0px 0px 0px 1px inset rgba(48, 114, 190, 1), 0 1px 3px rgba(0, 0, 0, 0.08)',
    '&:hover': {
      borderColor: 'rgba(48, 114, 190, 1)',
      cursor: 'pointer'
    }
  }),
  imgWrapper: (imageHasLoaded: boolean) =>
    css({
      width: '290px',
      height: `${imageHasLoaded ? 290 : 0}px`,
      position: 'relative',
      overflow: 'hidden'
    }),
  previewImg: css({
    height: '290px',
    minWidth: 'auto',
    position: 'absolute',
    top: '50%',
    left: '50%',
    transform: 'translate(-50%, -50%)'
  }),
  name: css({
    flex: '1 0 auto',
    fontWeight: 'bold',
    textTransform: 'capitalize'
  }),
  sku: css({
    flex: '0 1 auto',
    color: tokens.colorTextLight,
    fontSize: tokens.fontSizeS,
    marginTop: 0,
    marginBottom: 0
  }),
  skeletonImage: css({
    width: '100%',
    height: '290px'
  }),
  check: (isSelected: boolean) =>
    css({
      opacity: isSelected ? 1 : 0,
      position: 'absolute',
      top: tokens.spacingL,
      right: tokens.spacingL,
      transition: `opacity ${tokens.transitionDurationDefault} ${tokens.transitionEasingDefault}`
    })
};

export const ProductListItem = (props: Props) => {
  const { product, isSelected, selectProduct } = props;
  const [imageHasLoaded, setImageHasLoaded] = useState(false);

  return (
    <div className={styles.productWrapper}>
      <div
        role="switch"
        aria-checked={isSelected}
        tabIndex={-1}
        className={`${styles.product} ${isSelected ? styles.selectedProduct : ''}`}
        onKeyUp={noop}
        onClick={() => selectProduct(product.sku)}>
        {!imageHasLoaded && (
          <SkeletonContainer className={styles.skeletonImage}>
            <SkeletonImage width={400} height={290} />
          </SkeletonContainer>
        )}
        <div className={styles.imgWrapper(imageHasLoaded)}>
          <img
            onLoad={() => setImageHasLoaded(true)}
            style={{ display: imageHasLoaded ? 'block' : 'none' }}
            src={product.image}
            alt="product preview"
            className={styles.previewImg}
            data-test-id="image"
          />
          <img className={styles.check(isSelected)} src={activeProductCheck} alt="check" />
        </div>
        <p className={styles.name}>{product.name}</p>
        <p className={styles.sku}>{product.sku}</p>
      </div>
    </div>
  );
};
