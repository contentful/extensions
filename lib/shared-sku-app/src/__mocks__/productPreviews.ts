import { Product } from '../interfaces';

const productPreviews: Product[] = [
  {
    id: '28a7442d',
    image: 'https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-1.jpg',
    name: 'Dress Twin-Set rose',
    sku: 'M0E20130820E90Z',
    externalLink: 'https://mc.commercetools.com/example-project/products/28a7442d/general'
  },
  {
    id: '4e5446e8',
    image: 'https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-2.jpg',
    name: 'Michael Kors – Shopper “Jet Set Travel”',
    sku: 'A0E2300FX102203',
    externalLink: 'https://mc.commercetools.com/example-project/products/4e5446e8/general'
  },
  {
    id: 'db3c234l',
    image: 'https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-3.jpg',
    name: 'Shirt “Jenny“ Polo Ralph Lauren white',
    sku: 'M0E21300900DZN7',
    externalLink: 'https://mc.commercetools.com/example-project/products/db3c234l/general'
  }
];

export default productPreviews;
