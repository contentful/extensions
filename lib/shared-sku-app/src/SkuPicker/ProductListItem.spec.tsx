import React from 'react';
import { configure, render, cleanup } from '@testing-library/react';
import { Props, ProductListItem } from './ProductListItem';
import productPreviews from './__mocks__/productPreviews';

configure({
  testIdAttribute: 'data-test-id'
});

const defaultProps: Props = {
  product: productPreviews[0],
  selectProduct: jest.fn(),
  isSelected: false
};

const renderComponent = (props: Props) => {
  return render(<ProductListItem {...props} />);
};

describe('ProductListItem', () => {
  afterEach(cleanup);

  it('should render successfully', async () => {
    const component = renderComponent(defaultProps);
    expect(component.container).toMatchSnapshot();
  });

  it('should render successfully the isSelected variation', () => {
    const component = renderComponent({ ...defaultProps, isSelected: true });
    expect(component.container).toMatchSnapshot();
  });
});
