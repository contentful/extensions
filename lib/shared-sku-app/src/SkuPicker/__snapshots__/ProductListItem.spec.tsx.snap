// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductListItem should render successfully 1`] = `
<div>
  <div
    class="css-rjsqim"
  >
    <div
      aria-checked="false"
      class="css-1n7zoyo "
      role="switch"
      tabindex="-1"
    >
      <img
        alt="product preview"
        class="css-8atqhb"
        data-test-id="image"
        src="https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-1.jpg"
        style="display: block;"
      />
      <p
        class="css-182natc"
      >
        Dress Twin-Set rose
      </p>
      <p
        class="css-13zy297"
      >
        M0E20130820E90Z
      </p>
    </div>
  </div>
</div>
`;

exports[`ProductListItem should render successfully the isSelected variation 1`] = `
<div>
  <div
    class="css-rjsqim"
  >
    <div
      aria-checked="true"
      class="css-1n7zoyo css-1rtg1va"
      role="switch"
      tabindex="-1"
    >
      <img
        alt="product preview"
        class="css-8atqhb"
        data-test-id="image"
        src="https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-1.jpg"
        style="display: block;"
      />
      <p
        class="css-182natc"
      >
        Dress Twin-Set rose
      </p>
      <p
        class="css-13zy297"
      >
        M0E20130820E90Z
      </p>
    </div>
  </div>
</div>
`;

exports[`ProductListItem should render successfully the loading variation 1`] = `
<div>
  <div
    class="css-rjsqim"
  >
    <div
      aria-checked="false"
      class="css-1n7zoyo "
      role="switch"
      tabindex="-1"
    >
      <svg
        aria-label="Loading component..."
        class="SkeletonContainer__SkeletonContainer___23jiu css-2z7te8"
        data-test-id="cf-ui-skeleton-form"
        height="100%"
        preserveAspectRatio="xMidYMid meet"
        role="img"
        width="100%"
      >
        <title>
          Loading component...
        </title>
        <rect
          clip-path="url(#cf-ui-skeleton-clip-id)"
          height="100%"
          style="fill: url(#cf-ui-skeleton-clip-gradient);"
          width="100%"
          x="0"
          y="0"
        />
        <defs>
          <clippath
            id="cf-ui-skeleton-clip-id"
          >
            <rect
              height="400"
              rx="0"
              ry="0"
              width="400"
            />
          </clippath>
          <lineargradient
            id="cf-ui-skeleton-clip-gradient"
          >
            <stop
              offset="0%"
              stop-color="#e5ebed"
              stop-opacity="1"
            >
              <animate
                attributeName="stop-color"
                dur="2s"
                id="animation1"
                repeatCount="indefinite"
                values="#e5ebed; #f7f9fa; #e5ebed"
              />
            </stop>
            <stop
              offset="50%"
              stop-color="#f7f9fa"
              stop-opacity="1"
            >
              <animate
                attributeName="stop-color"
                begin="animation1.begin+0.25s"
                dur="2s"
                repeatCount="indefinite"
                values="#e5ebed; #f7f9fa; #e5ebed"
              />
            </stop>
            <stop
              offset="100%"
              stop-color="#e5ebed"
              stop-opacity="1"
            >
              <animate
                attributeName="stop-color"
                begin="animation1.begin+0.5s"
                dur="2s"
                repeatCount="indefinite"
                values="#e5ebed; #f7f9fa; #e5ebed"
              />
            </stop>
          </lineargradient>
        </defs>
      </svg>
      <img
        alt="product preview"
        class="css-8atqhb"
        data-test-id="image"
        src="https://s3-eu-west-1.amazonaws.com/commercetools-example/products/picture-1.jpg"
        style="display: none;"
      />
      <p
        class="css-182natc"
      >
        Dress Twin-Set rose
      </p>
      <p
        class="css-13zy297"
      >
        M0E20130820E90Z
      </p>
    </div>
  </div>
</div>
`;
